### DISCUSSION TITLE
Implementing a squarewave: Can’t take remainder of InfiniteOpt.GeneralVariableRef

### DISCUSSION THREAD

**Discussion by codercahol** [infiniteopt/InfiniteOpt.jl#337] [Q&A] [answered]
I have a squarewave implemented like so:

```
function sqw(t; period::Float64=1, duty::Float64=0.5)
	return t % period < duty*period ? 1.0 : 0.0
end
```

I am running an InfiniteOpt.jl optimization where I pass in `sqw` as a parameter and call it with an infinite parameter:
```
@infinite_parameter(model, t ∈ [0, T], num_supports = n)
@constraint(model, ∂(x, t) == sqw(t))
```

I get the error 
```
MethodError: no method matching rem(::InfiniteOpt.GeneralVariableRef, ::Float64)
```

Clearly, the `rem` has not been implemented for InfiniteOpt

So I have two options:
1) find another way to model the signal
2) implement `rem` for the `GeneralVariableRef`

If #1 is do-able in a InfiniteOpt.jl compatible way, I would love to hear suggestions.

Otherwise, how can I get started on #2?

P.S. - [cross-posted it on the Julia forums](https://discourse.julialang.org/t/implementing-a-squarewave-cant-take-remainder-of-infiniteopt-generalvariableref/111955/1)

**Comment by pulsipher** [comment, marked as answer]
Hi @codercahol, sorry for the late response (I missed the notification for this post).

Like JuMP, InfiniteOpt is a symbolic optimization language where expressions like `t % period < duty*period ? 1.0 : 0.0` are not readily supported (see https://jump.dev/JuMP.jl/stable/manual/nonlinear/#Supported-operators).

However, since your `sqw` function only depends on an infinite parameter, InfiniteOpt provides a way around this limitation via the use of a modeling object called a [parameter function](https://infiniteopt.github.io/InfiniteOpt.jl/stable/guide/expression/#par_func_docs). Parameter functions provide a way to embed an arbitrary Julia function the takes infinite parameter(s) as input and outputs a single number.

In your case we can write:
```julia
function sqw(t; period::Float64=1, duty::Float64=0.5)
	return t % period < duty*period ? 1.0 : 0.0
end

model = InfiniteModel()
@infinite_parameter(model, t in [0 , T], num_supports = n)

# create a parameter function called wrapped_sqw that uses sqw with infinite parameter t as an input
@parameter_function(model, wrapped_sqw == sqw(t))
@constraint(model, ∂(x, t) == wrapped_sqw)
```

An alternative, but equivalent syntax is:
```julia
function sqw(t; period::Float64=1, duty::Float64=0.5)
	return t % period < duty*period ? 1.0 : 0.0
end

model = InfiniteModel()
@infinite_parameter(model, t in [0 , T], num_supports = n)

# use parameter_function to embed sqw(t) within the constraint expression
@constraint(model, ∂(x, t) == parameter_function(sqw, t))
```

**Reply by codercahol** [reply]
Thanks!
